runtime-modules/content-directory/src/lib.rs

```diff
@@ -1198,37 +1268,62 @@ decl_module! {
             property_values: BTreeMap<PropertyId, PropertyValue<T>>
         ) -> dispatch::Result {
 
+            // Retrieve Class, Entity and ensure given have access to the Entity under given entity_id
             let (class, entity, _) = Self::ensure_class_entity_and_access_level(origin, entity_id, &actor)?;
 
+            // Ensure Schema under given id is not added to given Entity yet
             entity.ensure_schema_id_is_not_added(schema_id)?;
 
-            let schema = class.ensure_schema_exists(schema_id)?;
+            // Ensure provided property_values are not added to the Entity values map yet
+            entity.ensure_property_values_are_not_added(&property_values)?;
+
+            // Ensure Class Schema under given index exist, return corresponding Schema
+            let schema = class.ensure_schema_exists(schema_id)?.to_owned();
 
+            // Ensure provided schema can be added to the Entity
             schema.ensure_is_active()?;
 
-            let entity_values = entity.get_values_ref();
+            // Ensure all provided property values are for properties in the given schema
+            schema.ensure_has_properties(&property_values)?;
 
-            // Updated entity values, after new schema support added
-            let mut entity_values_updated = entity.values.clone();
+            // Ensure all property_values under given Schema property ids are valid
+            let entity_controller = entity.get_permissions_ref().get_controller();
 
-            // Entities, which rc should be incremented
-            let mut entity_ids_to_increase_rcs = EntitiesRc::<T>::default();
+            let class_properties = class.properties;
 
-            for prop_id in schema.get_properties().iter() {
-                if entity_values.contains_key(prop_id) {
-                    // A property is already added to the entity and cannot be updated
-                    // while adding a schema support to this entity.
-                    continue;
-                }
+            // Perform all required checks to ensure property_values under provided schema property_ids are valid
+            Self::ensure_property_values_are_valid(
+                &class_properties, schema.get_properties(), entity_controller, &property_values,
```

Comment:
The whole point of `ValuesForExistingProperties` was to avoid what is still being done here in this routine.

Please make sure to use it everywhere it may be of use, which is not only for this method in many other extrinsics, but probably also other unrelated methods.

Reply:
Why do you need to pass `schema.get_properties()`, even without the `ValuesForExistingProperties` change?

At this point, we have ensured that `property_values` only has values for properties in the schema, why do we need to refer back to it as a separate input here?

Reply:
Why does this method do lots of combined checks for validity of new values, but this is its own standalone check:
`ensure_property_values_unique_option_satisfied`? It seems that if there is one validity call should cover them all, however the call should of course do them in separate steps for clarity, as you already do in `ensure_property_value_to_update_is_valid`.

Reply:
Good point, already decoupled required check as well here affc8b4